### DISCUSSION TITLE
Chat

### DISCUSSION THREAD

**Discussion by kmyk** [kmyk/Jikka#24] [General]
This is a chat space. Both English and Japanese are welcome. Please use like a Twitter.

**Comment by kmyk** [comment]
ここの chat 欄をもっと雑に使っていくといい気がする

**Reply by kmyk** [reply]
よし

**Reply by kmyk** [reply]
にゃん

**Comment by kmyk** [comment]
thread 式なのか。話題ごとに整理するとかじゃなくて、もっと Twitter 感覚で適当に書き込める感じがいいんだけどな

**Reply by kmyk** [reply]
まあいいか。Twitter だと思えば任意の場所が Twitter です

**Comment by kmyk** [comment]
最適化の実装を任せられるぽいので、私は人には任せられないしんどいパートをやっていきます (しんどい…)

**Reply by kmyk** [reply]
とりあえず core 言語のパーサを書きます

**Reply by kmyk** [reply]
なんと C++ のパーサの予定もあるぞ！ なんで 1 個のコンパイラに 3 個のパーサがいるんや、おかしいやろ

**Reply by kmyk** [reply]
Python はしんどかったが core なら単純だからいける気がしてきた

**Reply by kmyk** [reply]
できた

**Reply by kmyk** [reply]
バグってる

**Comment by kmyk** [comment]
@hotman78 chat 欄がここにあります。質問があるときとかないときとかに気軽に使ってね

**Reply by hotman78** [reply]
ありがとうございます！

**Comment by hotman78** [comment]
二度手間防止/ルール案の発見のためにoptimization.mdに書いてないような細かい最適化を挙げてく

**Reply by kmyk** [reply]
とりあえず、細かい最適化のうちで他のルールを書くときに重要そうなやつを挙げておきます

- ArithmeticalExpr: 算術式を整理する。`n - (n - i - 1) - 1` が `i` になったりする。式の中の自由変数が減るとやれる最適化が増える
- Alpha: alpha 変換をする。`(fun x -> fun x -> e(x))` みたいな式の変数名を付け替えて被りをなくし `(fun x -> fun y -> e(y))` にする
- Beta: beta 簡約をする。つまり `(fun x -> e(x)) a` を `e(a)` にする
- BubbleLet: `foldl (fun y x -> let z = c in e) y0 xs` を可能なら `let z = c in foldl (fun y x -> e) y0 xs` にする。`c` の計算が重いなら計算量が落ちる
- ConstantFolding: すべてが定数の式を計算してしまう。たとえば `3 + 4` を `7` に置き換えるとか
- Eta: eta 変換をする。たとえば `foldl (+) 0 xs` を `foldl (fun y x -> y + x) 0 xs` にする

完全なリストは https://kmyk.github.io/Jikka/haddock/ を見るのがよい気がする

**Comment by hotman78** [comment]
バグの内容メモ
coreにするところでバグってる

バグる式
```python
# before
return (-n)*(n+1)
# after
return -n+n*n ;

# before
return (-1)*(n+1)
# after
return -n+2

# before
return 2*(n+1)
# after
return n * 2 + 4

# before
return 2*(n-1)
# after
return n * 2 - 4


```

バグらない式
```python
# before
return 2*(n-n*n)

# after
return 2*n-2*n*n

# before
return 2*(1+3)

# after
return 8

```

**Reply by hotman78** [reply]
https://github.com/kmyk/Jikka/blob/8aac3c02a5d035df06d5024eca3e30ee5e017536/src/Jikka/Core/Convert/StrengthReduction.hs#L44
ここ怪しいけどここだけではなさそう

**Reply by hotman78** [reply]
定数項が二倍になってる？

**Comment by kmyk** [comment]
rewrite rule が今より簡単に書けるようになりました。https://github.com/kmyk/Jikka/pull/133

いままでは

``` haskell
simpleRewriteRule $ \case
  Map' t1 t2 f (Reversed' _ xs) -> return' $ Reversed' t2 (Map' t1 t2 f xs)
  _ -> Nothing
```

と書いてた部分が、代わりに

``` haskell
[r| "map/reversed" forall f xs. map f (reversed xs) = reversed (map f xs) |]
```

とも書けます。簡潔で見やすくなって、書きやすい + どういう rule があるのか分かりやすいのでうれしい。


Template Haskell という、Rust で言うところのマクロみたいな機能を使って実装されています。

**Comment by kmyk** [comment]
rule のバグが後になってから発見されるやつしんどいし、もっとテストをたくさん用意しておかないとだめそう。ABC の B, C あたりをひたすら解いて examples/ の中に突っ込む会とかが必要そう

**Reply by hotman78** [reply]
一気になるとしんどそうなので、合間合間でやっていこうと思います

**Comment by kmyk** [comment]
[docs/internal.ja.md](https://github.com/kmyk/Jikka/blob/master/docs/internal.ja.md) を要約して主張を足した感じの記事を書きました

プログラミング言語処理系の作り方 (競プロer向け) - うさぎ小屋
https://kimiyuki.net/blog/2021/08/05/how-to-make-your-own-interpreter/

**Comment by kmyk** [comment]
@hotman78 #151 #152 これらに興味ありませんか？ <O(N), O(1)> GCD チャンスです

**Reply by hotman78** [reply]
ちょうど何から手を付けるか悩んでたのでやります！
issueにないdata_structureも追加できそうなものはどんどん挙げていいですかね？

**Reply by kmyk** [reply]
> 追加できそうなものはどんどん挙げていいですかね？

はい。貼ってうれしそうなものはなんでも歓迎です。思い付いたら新しい issue として立てておくとよい

**Comment by hotman78** [comment]
むーん
min/maxとgcd/lcmで文法が異なるの若干めんどうの元になりそうな気がしてます
(現状変換されなくて悲しくなってるだけなのですが...)

**Reply by kmyk** [reply]
「文法が異なる」とは、min/max だと `max2 : a -> a -> a` と `max1 : a list -> a` があるけど gcd/lcm だと `gcd : int -> int -> int` しかないということでしょうか？
たしかにそれは面倒そう。`gcd1 : int list -> int` とかを足してしまってもよいかもしれません。

新しい組み込み関数として足してもよいですし、foldl を使って alias として書くこともできます。以下のように書けば `foldl gcd 0` の alias として `gcd1` が使えるようになります。

``` haskell
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE PatternSynonyms #-}
{-# LANGUAGE ViewPatterns #-}

pattern Gcd1 e <-
  (\case Foldl' IntTy IntTy (Lam2 y IntTy x IntTy (Gcd' (Var y') (Var x'))) (LitInt' 0) e | y' == y && x' == x -> Just e; _ -> Nothing -> Just e)
  where
    Gcd1 e = Foldl' IntTy IntTy (Lam2 "y" IntTy "x" IntTy (Gcd' (Var "y") (Var "x"))) (LitInt' 0) e
```

**Reply by hotman78** [reply]
patternなるほどです...
新しい組み込み関数として足してしまいました
上で考えていた内容は少し違う内容だったのですが、自己解決しました
Lit (LitBuiltin Gcd [])
Lit (LitBuiltin Min2 [IntTy])
が同じ型になってくれるので大丈夫そうでした

**Comment by kmyk** [comment]
整数に最大値はないので max nil がエラーになるやつ、どうにかならないかな。「どうせ後で `int64_t` になるので、最大値はあります！w」ってやっちゃだめかな？ 変なバグり方してやばそうではあるが

**Comment by kmyk** [comment]
ここの存在、普通に考えて誰も気付いてないと思うので ping しておく @uta8a  @riantkb
まあ Twitter とか issue とかでいいという話もあるんですが

**Comment by kmyk** [comment]
core から C++ への変換を「とりあえず変換する。そして無駄なコピーを消す」できると思うじゃん。ふたつのパートを分けておいた方が見通しよく書けるはずだったんですよ。なんだかそれではだめらしいです。つらすぎ

**Comment by uta8a** [comment]
examplesを追加する作業で、examples/README.mdにはなんて書けばいいんだろうか。wipのABC-Bは特別にすごいことをしているわけではないので、Real tasks of competitive programmingに軽く書いちゃっていいのかな？と思いました

**Reply by uta8a** [reply]
これ慣れてないから非本質で悩んでいるだけな気がしてきた。気にしないことにする。

**Reply by kmyk** [reply]
気にせず軽く書いちゃって大丈夫です
「愚直がそのまま通る」というのが分かるようになってればうれしいけど、まあ ABC の B だしそのことは問題 ID からも分かる

**Comment by riantkb** [comment]
ループ内の割り算が消えるの、コメントアウトデバッグをしたら `reduceFold` が悪さしてそう？ ということが分かったが、何もわからず
https://github.com/kmyk/Jikka/blob/45c04f382c761e0716041efc284b566a6a7d2ca1/src/Jikka/Core/Convert/ShortCutFusion.hs#L116-L119

**Reply by kmyk** [reply]
`reduceFold` は正しそうに見えます。となると `reduceFold` が動作することで起動されるやつが怪しくて、でも検索 https://github.com/kmyk/Jikka/search?q=Iterate%27 してみたけどそれっぽいのを見つけられない

**Reply by riantkb** [reply]
`MatrixExponentiation` だけをコメントアウトしても正しいコードが得られましたね……（というかコードを読んだら完全にそれっぽいことをしているやつだ）

**Comment by uta8a** [comment]
連絡: haskellのフォーマッタormoluのtrailing newline問題、すぐに向こうの方が直してくれた。ありがたい
https://github.com/tweag/ormolu/pull/744

**Comment by kmyk** [comment]
core から C++ への変換をずっと考えています。やればできるはずなんですけど、頭がこわれる
https://github.com/kmyk/Jikka/issues/177

**Reply by kmyk** [reply]
https://github.com/kmyk/Jikka/pull/168 できた :tada:

**Comment by kmyk** [comment]
@hotman78 そういえば FFT のやつとか調子どうですか (何を実装するにしてもしんどすぎることについては申し訳なく思っています……)

**Reply by hotman78** [reply]
別の用事が重なって途切れてから全然手をつけられてませんでした...
落ち着いてきたので再開します！！

**Comment by kmyk** [comment]
@soraiemame ここの存在に気付いてないかもなので念のため ping しておく

**Comment by kmyk** [comment]
@各位
「Jikka の実装のここが分かりにくい」みたいなのがあったら教えてください。もうすでに自己解決したものでもよいです。ドキュメントに書くので

**Comment by kmyk** [comment]
YouTube で開発配信したりとか Discord で画面を共有してペアプロしたりとかしたらいい気がしてきた
